manager/usermanager/createUser.go

```diff
@@ -68,16 +68,30 @@ func CreateUser(userStore *store.UserStore, user *models.UserCredentials, isSign
 
 // CreateSocialUser creates a user if the user opts logging in with some oauth
 func CreateSocialUser(userStore *store.UserStore, user *models.UserCredentials) error {
-	query := bson.M{"email": user.Email, "kind": models.LocalAuth}
-	storedUser, err := userStore.GetUser(query)
-	if err != nil && err == mgo.ErrNotFound {
-		user.UserName = generateUserName(user.Name)
-		user.UID = uuid.Must(uuid.NewRandom()).String()
-	} else if err != nil {
+	userWithSameEmail, err := GetUser(userStore, bson.M{"email": user.Email})
+	if err == nil && userWithSameEmail != nil {
+		// If a user with this email is already existing then return error
+		return errors.ErrUserExists
+	} else if err != errors.ErrInvalidUser {
+		// If some error occurs other than invalid user (here invalid user
+		// means such a user does not exist)
 		return err
-	} else {
-		user.UserName = storedUser.UserName
-		user.UID = storedUser.UID
 	}
+
+	// If user with the given email does not exist.
+	// This infinite loop generates a username and checks whether this username
```

Comment:
What does this generation logic do? Why we didn't need a loop earlier?

Reply:
The logic is as follows -
- The loop will keep generating usernames and check them in the database if it is existing or not.
- As soon as it finds a username which is not existing in database the loop breaks.

Previously we were generating username and were assigning it directly as there is a random string appended for generating the username. Although there is a very rare chance for those random strings to be similar but still checking it once in the database makes more sensefor any edge case. So the loop comes in need for this edge case

Reply:
Any reason why we cannot use the username from the social-auth?

Reply:
Also, its a good practice to have a mini sleep in an infinite for loop, else CPU can go for a spin. In this case, a sleep of 100ms should suffice.

Reply:
Then two different users may have the same username. E.g A user have a github username as `user123` and a different user can have his google username as `user123`.

Reply:
How will that cause a problem?

Reply:
Typically, we need not worry about uniqueness of username fields. Today, we only worry about email.
But, for the time being, you can add a mini sleep as per this comment.
https://github.com/mayadata-io/kubera-auth/pull/59#discussion_r609578423

Reply:
Lets say if first user signs up through github he will be assigned the username `user123` and the second user signs up through google he will also e assigned the username `user123`. So now 2 different users will have the same username in our database.
Although there `auth_type` will be different but many queries in the UI are just based on `username` only  as UI assumes them to be unique. So it will be causing problems.

Reply:
https://github.com/mayadata-io/kubera-auth/pull/59#discussion_r609584126 
Okay

Reply:
> Lets say if first user signs up through github he will be assigned the username `user123` and the second user signs up through google he will also e assigned the username `user123`. So now 2 different users will have the same username in our database.
> Although there `auth_type` will be different but many queries in the UI are just based on `username` only as UI assumes them to be unique. So it will be causing problems.

Okay, I see.

Reply:
Though I'm not sure, why UI expects username to be unique. For uniqueness purposes, we always had uuid and now email as well.

Reply:
Though we have changed queries from usernae based to uid based still there are many places in which username is being used. First we need to refactor that